File: WebFunc-node20-langgraph_pure_js/src/src/agent.js
import { ChatOpenAI } from "@langchain/openai";
import { SystemMessage } from "@langchain/core/messages";
import {
  Command,
  StateGraph,
  END,
  START,
  MemorySaver,
} from "@langchain/langgraph";
import { ClientStateAnnotation } from "@cloudbase/agent-adapter-langgraph";

// MemorySaver: 内存级对话历史存储，支持多轮对话上下文
// 生产环境可替换为持久化存储（如数据库）
const checkpointer = new MemorySaver();

/**
 * 聊天节点 - LangGraph 工作流的核心节点
 *
 * 负责：
 * 1. 初始化大模型并绑定客户端工具
 * 2. 构造系统提示词
 * 3. 调用大模型生成回复
 * 4. 返回状态更新命令
 *
 * @param {object} state - 当前工作流状态
 * @param {object} config - 运行配置
 * @returns {Command} 状态更新命令
 */
async function chatNode(state, config) {
  // 配置 OpenAI 兼容的大模型
  const model = new ChatOpenAI({
    model: process.env.OPENAI_MODEL,
    apiKey: process.env.OPENAI_API_KEY,
    configuration: {
      baseURL: process.env.OPENAI_BASE_URL,
      defaultHeaders: {
        Accept: "text/event-stream",
      },
    },
  });

  // 设置默认递归限制，防止无限循环
  const effectiveConfig = config ?? { recursionLimit: 25 };

  // 绑定客户端传入的工具，禁用并行工具调用以保证执行顺序
  const modelWithTools = model.bindTools([...(state.client?.tools || [])], {
    parallel_tool_calls: false,
  });

  // 系统提示词：定义 Agent 的角色和行为
  const systemMessage = new SystemMessage({
    content: "你是一位精通云开发 CloudBase 的专家，擅长回答任何相关的问题。",
  });

  // 调用大模型，传入系统消息和历史对话
  const response = await modelWithTools.invoke(
    [systemMessage, ...state.messages],
    effectiveConfig,
  );

  // 返回 Command，更新消息列表并结束工作流
  return new Command({
    goto: END,
    update: {
      messages: [response],
    },
  });
}

/**
 * 构建 LangGraph 工作流
 *
 * 工作流结构：
 *   START -> chat_node -> END
 *
 * 这是一个简单的单节点工作流，可扩展为多节点复杂流程
 */
const workflow = new StateGraph(ClientStateAnnotation)
  .addNode("chat_node", chatNode)
  .addEdge(START, "chat_node")
  .addEdge("chat_node", END);

/**
 * 创建 LangGraph 工作流实例
 * 这是底层的 Agent 逻辑，负责：
 * - 与大模型交互
 * - 管理对话历史
 * - 处理工具调用
 */
export function createAgenticChatGraph() {
  return workflow.compile({
    checkpointer,
  });
}

File: WebFunc-python310-langgraph/src/agent.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Agentic Chat Agent Implementation.

This module provides a complete implementation of a conversational AI agent
using LangGraph workflows and OpenAI-compatible language models. It demonstrates
how to build streaming chat agents with proper state management, message
handling, and error recovery for production use.

Key Features:
    - Real-time streaming response generation
    - Conversation history management with proper role mapping
    - Integration with OpenAI-compatible API endpoints
    - Memory persistence using LangGraph checkpointers
    - Tool binding and function calling support
    - Robust error handling and graceful degradation
    - Environment-based configuration management

Architecture:
    The agent uses a simple linear workflow (START -> ai_response -> END)
    with state management for conversation continuity. It supports various
    message types (Human, System, AI) and provides streaming output for
    better user experience.

Usage:
    This agent can be deployed as a web service using the AG-Kit server
    framework or used programmatically for building conversational applications.
"""

import json
import os
from typing import Any, List, Optional
from dotenv import load_dotenv
load_dotenv()

from langchain_core.messages import AIMessage, SystemMessage, convert_to_openai_messages
from langchain_core.messages.ai import add_ai_message_chunks
from langchain_core.runnables import RunnableConfig
from langchain_core.utils.function_calling import convert_to_openai_function
from langchain_openai import ChatOpenAI

from langgraph.checkpoint.memory import MemorySaver
from langgraph.graph import END, START, MessagesState, StateGraph
from langgraph.graph.state import CompiledStateGraph


class State(MessagesState):
    tools: List[Any]


def chat_node(state: State, config: Optional[RunnableConfig] = None) -> dict:
    try:
        # Validate required environment variables
        model_name = os.getenv("OPENAI_MODEL")
        api_key = os.getenv("OPENAI_API_KEY")
        base_url = os.getenv("OPENAI_BASE_URL")
        
        if not model_name:
            raise ValueError("OPENAI_MODEL environment variable is not set")
        if not api_key:
            raise ValueError("OPENAI_API_KEY environment variable is not set")
        if not base_url:
            raise ValueError("OPENAI_BASE_URL environment variable is not set")
        
        # Create LangChain ChatOpenAI model from environment variables
        chat_model = ChatOpenAI(
            model=model_name,
            api_key=api_key,
            base_url=base_url,
        )

        # Set default config if none provided
        if config is None:
            config = RunnableConfig(recursion_limit=25)

        # Process and bind tools if available
        tools = state.get("tools", [])
        if tools:
            # Convert tools to OpenAI format, supporting both string and dict parameters
            tool_list = []
            for tool in tools:
                params = tool["parameters"]
                # Support both JSON string and dict format for flexibility
                if isinstance(params, str):
                    params = json.loads(params)
                tool_list.append({**tool, "parameters": params})
            
            tools_list = [convert_to_openai_function(item) for item in tool_list]
            chat_model_with_tools = chat_model.bind_tools(tools_list)
        else:
            chat_model_with_tools = chat_model

        # Create system message to set assistant behavior
        system_message = SystemMessage(content="You are a helpful assistant.")
        messages = [system_message, *convert_to_openai_messages(state["messages"])]

        # Stream through LangChain's standard streaming mechanism
        # The streaming will be automatically captured by LangGraph and AG-Kit
        chunks = []
        for chunk in chat_model_with_tools.stream(messages, config):
            chunks.append(chunk)  # Collect for final message

        # Merge all chunks into complete AIMessage
        if chunks:
            from langchain_core.messages import AIMessageChunk

            # Convert chunks to AIMessageChunk if needed
            ai_chunks = [
                chunk if isinstance(chunk, AIMessageChunk) else AIMessageChunk(content=str(chunk)) for chunk in chunks
            ]
            merged_message = add_ai_message_chunks(*ai_chunks)
            return {"messages": [merged_message]}
        else:
            return {"messages": []}
    except Exception as e:
        # Handle errors gracefully with informative messages
        return {"messages": [AIMessage(content=f"Error: {str(e)}")]}


def build_agentic_chat_workflow() -> CompiledStateGraph:
    # Create state graph for the chat workflow
    graph = StateGraph(State)
    memory = MemorySaver()  # In-memory checkpointer for conversation persistence

    # Add the AI response generation node
    graph.add_node("ai_response", chat_node)
    graph.set_entry_point("ai_response")

    # Define simple linear workflow: START -> ai_response -> END
    graph.add_edge(START, "ai_response")
    graph.add_edge("ai_response", END)

    # Compile workflow with memory checkpointer for conversation history
    return graph.compile(checkpointer=memory)
